File: go/internal/harness/procfs.go
package harness

import (
	"bufio"
	"os"
	"strconv"
	"strings"
	"syscall"
)

// memStatus reads VmRSS and VmHWM (in KiB) from /proc/self/status.
func memStatus() (rssKB, hwmKB int64) {
	f, err := os.Open("/proc/self/status")
	if err != nil {
		return -1, -1
	}
	defer f.Close()
	rssKB, hwmKB = -1, -1
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := sc.Text()
		switch {
		case strings.HasPrefix(line, "VmRSS:"):
			rssKB = parseKB(line)
		case strings.HasPrefix(line, "VmHWM:"):
			hwmKB = parseKB(line)
		}
	}
	return rssKB, hwmKB
}

func parseKB(line string) int64 {
	fields := strings.Fields(line)
	if len(fields) < 2 {
		return -1
	}
	n, err := strconv.ParseInt(fields[1], 10, 64)
	if err != nil {
		return -1
	}
	return n
}

// ThreadCount returns the number of OS threads of this process.
func ThreadCount() int64 {
	b, err := os.ReadFile("/proc/self/status")
	if err != nil {
		return -1
	}
	for _, line := range strings.Split(string(b), "\n") {
		if strings.HasPrefix(line, "Threads:") {
			return parseKB(line)
		}
	}
	return -1
}

type rusage struct {
	utimeNs, stimeNs, minflt, majflt, nvcsw, nivcsw int64
}

func getRusage() rusage {
	var ru syscall.Rusage
	if err := syscall.Getrusage(syscall.RUSAGE_SELF, &ru); err != nil {
		return rusage{}
	}
	return rusage{
		utimeNs: ru.Utime.Nano(),
		stimeNs: ru.Stime.Nano(),
		minflt:  ru.Minflt,
		majflt:  ru.Majflt,
		nvcsw:   ru.Nvcsw,
		nivcsw:  ru.Nivcsw,
	}
}

func (a rusage) delta(b rusage) map[string]int64 {
	return map[string]int64{
		"utime_ns": b.utimeNs - a.utimeNs,
		"stime_ns": b.stimeNs - a.stimeNs,
		"minflt":   b.minflt - a.minflt,
		"majflt":   b.majflt - a.majflt,
		"nvcsw":    b.nvcsw - a.nvcsw,
		"nivcsw":   b.nivcsw - a.nivcsw,
	}
}

File: rust/common/src/procfs.rs
//! Process statistics, read exactly like go/internal/harness/procfs.go.

use std::fs;

fn status_field(text: &str, key: &str) -> i64 {
    for line in text.lines() {
        if let Some(rest) = line.strip_prefix(key) {
            return rest.split_whitespace().next().and_then(|v| v.parse().ok()).unwrap_or(-1);
        }
    }
    -1
}

/// (VmRSS, VmHWM) in KiB from /proc/self/status.
pub fn mem_status() -> (i64, i64) {
    match fs::read_to_string("/proc/self/status") {
        Ok(t) => (status_field(&t, "VmRSS:"), status_field(&t, "VmHWM:")),
        Err(_) => (-1, -1),
    }
}

/// Number of OS threads in this process.
pub fn thread_count() -> i64 {
    match fs::read_to_string("/proc/self/status") {
        Ok(t) => status_field(&t, "Threads:"),
        Err(_) => -1,
    }
}

#[derive(Clone, Copy, Default, Debug)]
pub struct Rusage {
    pub utime_ns: i64,
    pub stime_ns: i64,
    pub minflt: i64,
    pub majflt: i64,
    pub nvcsw: i64,
    pub nivcsw: i64,
}

pub fn rusage() -> Rusage {
    // SAFETY: getrusage only writes into the zero-initialised struct we own.
    unsafe {
        let mut ru: libc::rusage = std::mem::zeroed();
        if libc::getrusage(libc::RUSAGE_SELF, &mut ru) != 0 {
            return Rusage::default();
        }
        Rusage {
            utime_ns: ru.ru_utime.tv_sec * 1_000_000_000 + ru.ru_utime.tv_usec * 1_000,
            stime_ns: ru.ru_stime.tv_sec * 1_000_000_000 + ru.ru_stime.tv_usec * 1_000,
            minflt: ru.ru_minflt,
            majflt: ru.ru_majflt,
            nvcsw: ru.ru_nvcsw,
            nivcsw: ru.ru_nivcsw,
        }
    }
}

impl Rusage {
    pub fn delta_json(&self, b: &Rusage) -> serde_json::Value {
        serde_json::json!({
            "utime_ns": b.utime_ns - self.utime_ns,
            "stime_ns": b.stime_ns - self.stime_ns,
            "minflt": b.minflt - self.minflt,
            "majflt": b.majflt - self.majflt,
            "nvcsw": b.nvcsw - self.nvcsw,
            "nivcsw": b.nivcsw - self.nivcsw,
        })
    }
}
